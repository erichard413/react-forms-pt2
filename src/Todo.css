.Todo {
    text-align: center;
    margin-top: 5px;
    margin-left: auto;
    margin-right: auto;
    width: 500px;
    height: auto;
    background-color: rgb(183, 200, 216);
    border: 1px solid rgb(227, 233, 236);
    border-radius: 12px;
}

.Todo p {
    font-family:'Lucida Sans', 'Lucida Sans Regular', 'Lucida Grande', 'Lucida Sans Unicode', Geneva, Verdana, sans-serif
}
.Todo span {
    cursor: pointer;
}

.Todo-btns {
    padding-bottom: 10px;
}