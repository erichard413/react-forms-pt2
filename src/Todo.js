import React, {useState} from 'react';
import EditForm from './EditForm';
import './Todo.css';

const Todo = ({text, idx, removeTodo, markComplete, complete, editTodo}) => {
    
    const [isShown, setIsShown] = useState(false);

    const showForm = () => {
        if (isShown === true) {
            setIsShown(false);
        } else {
            setIsShown(true);
        }
    }


    return (
        <div className="Todo">
            <p style={complete ? {textDecoration: 'line-through'} : null}>{text}</p>
            <div className="Todo-btns">
            <span role="removeTodo" onClick={()=> removeTodo(idx)}>❌</span>
            <span style={{color: 'transparent', textShadow: '0 0 green'}} role="markComplete" onClick={()=> markComplete(idx)}>✔️</span>
            <span style={{color: 'transparent', textShadow: '0 0 grey'}} role="showForm" onClick={showForm}>✏️</span>
            </div>
            
            {isShown && <EditForm task={text} idx={idx} editTodo={editTodo} showForm={showForm} />}  
        </div>
    )
}

export default Todo;