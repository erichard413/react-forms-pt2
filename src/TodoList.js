import React, {useState} from 'react';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import {v4 as uuidv4} from 'uuid';
import './TodoList.css';



const TodoList = () => {
    
    const [todos, setTodos] = useState(()=> {
        const savedTodos = localStorage.getItem("todos");
        const initialValue = JSON.parse(savedTodos);
        return initialValue || "";
    });

    const removeTodo = (idx) => {
        setTodos(todos => {
            return todos.filter(t => t.idx !== idx);
        })
    }

    const markComplete = (idx) => {
        let copy = [...todos]
        let toDoCopy = copy.filter(t => t.idx === idx);
                toDoCopy[0].complete ? toDoCopy[0].complete = false : toDoCopy[0].complete = true;
        return setTodos(copy);
        }

    const addTodo = (text) => {
        let newTask = {
            text: text,
            complete: false,
            idx: uuidv4()
        }
        setTodos(todos => {
            let toDosCopy = [...todos, newTask]
            return toDosCopy
        })
    }

    const editTodo = (idx, text) => {
        setTodos(todos => {
            let todosCopy = [...todos]
            let toDoCopy = todosCopy.filter(t => t.idx === idx);
            toDoCopy[0].text = text;
            return [...todosCopy]
        })
    }

    localStorage.setItem("todos", JSON.stringify(todos));

    return (
        <div className="TodoList">
            <NewTodoForm addTodo={addTodo} />
            <h2>Current To Do's</h2>
            {todos.map(t=> (
                <Todo key={t.idx} text={t.text} idx={t.idx} removeTodo={removeTodo} complete={t.complete} markComplete={markComplete} editTodo={editTodo}/>
            ))}
        </div>
    )
}

export default TodoList;