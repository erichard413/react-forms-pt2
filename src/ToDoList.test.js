import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';
import Todo from './Todo';

it('should render OK', ()=>{
    render(<TodoList />)
})

it('should match snapshot', ()=> {
    const {asFragment} = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it('should start with no todos', ()=> {
    const result = render(<TodoList />);
    const todo = result.container.querySelector('.Todo');
    expect(todo).not.toBeInTheDocument();
});

it('should add new todo', ()=> {
    const result = render(<TodoList />);
    const {getByLabelText, queryByText} = result;
    const text = getByLabelText('Task');
    const btn = queryByText('Add Task');
    fireEvent.change(text, {target: {value: 'do test'}});
    fireEvent.click(btn);
    const todo = result.container.querySelector('.Todo');
    expect(todo).toBeInTheDocument();
});

it('should delete todo', ()=> {
    const result = render(<TodoList />);
    const {getByLabelText, queryByText, getByRole} = result;
    const text = getByLabelText('Task');
    const btn = queryByText('Add Task');
    fireEvent.change(text, {target: {value: 'do test'}});
    fireEvent.click(btn);
    const todo = result.container.querySelector('.Todo');

    const delBtn = getByRole('removeTodo');

    expect(todo).toBeInTheDocument();
    fireEvent.click(delBtn);
    expect(todo).not.toBeInTheDocument();
})