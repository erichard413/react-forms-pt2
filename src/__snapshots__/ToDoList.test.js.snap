// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form>
      <label
        for="text"
      >
        Task
      </label>
      <input
        id="text"
        name="text"
        placeholder="New Task"
        type="text"
        value=""
      />
      <button>
        Add Task
      </button>
    </form>
    <h2>
      Current To Do's
    </h2>
  </div>
</DocumentFragment>
`;
